sdlib/d/gc/page.d

```diff
@@ -796,22 +825,48 @@ private:
 					continue;
 				}
 
-				// If the cycle do not match, all the elements are dead.
-				if ((w & 0xff) != gcCycle) {
+				auto hasMeta = e.slabMetadataFlags.rawContent[0];
+
+				// If the cycle do not match, all the elements are dead. If no
+				// metadata exists, then we can safely clear the whole thing.
+				if (hasMeta == 0 && (w & 0xff) != gcCycle) {
 					deadExtents.insert(e);
 					continue;
 				}
 
 				auto oldOccupancy = e.slabData.rawContent[0];
 				auto newOccupancy = oldOccupancy & (w >> 8);
 
+				// need to check dying blocks for finalizers
+				auto toRemove = oldOccupancy ^ newOccupancy;
+				hasMeta &= toRemove;
+
+				if (hasMeta) {
+					import d.gc.slab;
+					auto ssize = binInfos[ec.sizeClass].slotSize;
+					while (hasMeta != 0) {
+						auto index = countTrailingZeros(hasMeta);
+						auto ptr = e.address + index * ssize;
+
+						auto metadata = SlotMetadata.fromBlock(ptr, ssize);
+						auto finalizer = metadata.finalizer;
+						if (finalizer) {
+							import d.finalizer;
+							__sd_run_finalizer(ptr, ssize - metadata.freeSpace,
+							                   finalizer);
+						}
+
+						hasMeta &= (hasMeta - 1);
+					}
+				}
```

Comment:
There is a lot of duplication in there. Is there really no way to factor this?

Reply:
I might be able to extract it to a common function.